File: v2/src/main/java/com/api/v2/customers/services/CustomerModificationServiceImpl.java
package com.api.v2.customers.services;

import com.api.v2.customers.domain.CustomerAuditTrail;
import com.api.v2.customers.domain.CustomerAuditTrailRepository;
import com.api.v2.customers.domain.CustomerRepository;
import com.api.v2.customers.dtos.CustomerResponseDto;
import com.api.v2.customers.utils.CustomerFinderUtil;
import com.api.v2.customers.utils.CustomerResponseMapperUtil;
import com.api.v2.persons.annotations.SSN;
import com.api.v2.persons.dtos.PersonModificationRequestDto;
import com.api.v2.persons.services.PersonModificationService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class CustomerModificationServiceImpl implements CustomerModificationService {

    @Autowired
    private CustomerFinderUtil customerFinderUtil;

    @Autowired
    private PersonModificationService personModificationService;

    @Autowired
    private CustomerAuditTrailRepository auditTrailRepository;

    @Autowired
    private CustomerRepository customerRepository;

    @Override
    public Mono<CustomerResponseDto> modify(@SSN String ssn, @Valid PersonModificationRequestDto requestDto) {
        return customerFinderUtil
                .find(ssn)
                .flatMap(existingCustomer -> auditTrailRepository.save(CustomerAuditTrail.of(existingCustomer))
                        .then(Mono.defer(() -> personModificationService.modify(existingCustomer.getPerson(), requestDto)))
                        .flatMap(person -> {
                            existingCustomer.setPerson(person);
                            return customerRepository.save(existingCustomer);
                        })
                        .flatMap(modifiedCustomer -> Mono.just(CustomerResponseMapperUtil.mapToDto(modifiedCustomer)))
                );
    }

}

File: kotlin/v3/src/main/kotlin/com/api/v3/customers/services/CustomerModificationServiceImpl.kt
package com.api.v3.customers.services

import com.api.v3.customers.domain.CustomerRepository
import com.api.v3.customers.utils.CustomerFinderUtil
import com.api.v3.persons.domain.PersonAuditTrail
import com.api.v3.persons.domain.PersonAuditTrailRepository
import com.api.v3.persons.domain.PersonRepository
import com.api.v3.persons.dtos.PersonModificationRequestDto
import jakarta.validation.Valid
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
internal class CustomerModificationServiceImpl: CustomerModificationService {

    @Autowired
    private lateinit var customerFinder: CustomerFinderUtil

    @Autowired
    private lateinit var customerRepository: CustomerRepository

    @Autowired
    private lateinit var personRepository: PersonRepository

    @Autowired
    private lateinit var personAuditTrailRepository: PersonAuditTrailRepository

    override suspend fun modify(ssn: String, requestDto: @Valid PersonModificationRequestDto) {
        return withContext(Dispatchers.IO) {
            val existingCustomer = customerFinder.find(ssn)
            val existingPerson = existingCustomer.person

            val auditTrail = PersonAuditTrail.of(existingPerson)
            var savedAuditTrail = personAuditTrailRepository.save(auditTrail)

            existingPerson.modify(requestDto)
            val savedPerson = personRepository.save(existingPerson)
            existingCustomer.person = savedPerson
            customerRepository.save(existingCustomer)
        }
    }

}